### ISSUE TITLE
Form construction during component lifecycle sometimes breaks

### ISSUE THREAD

**Issue by xealot** [angular/angular#12381]
**I'm submitting a ...** 
```
[x] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**
When inputs change in a component and you create the form again using `ngOnChanges`, custom ControlValueAccessors can break.

**Expected behavior**
Should not break.

**Minimal reproduction of the problem with instructions**
Enter random value in input box, hit `Change Form` button, enter another random value in the input box.

https://plnkr.co/edit/T6k3rClJW5dWU3yiaTpi?p=preview

**Please tell us about your environment:**
WebStorm, macOS, SystemJS

* **Angular version:** 2.1.0
* **Browser:** Probably All, but for sure Chrome
 * **Language:** all

**Comment by DzmitryShylovich** [comment]
why do u need to reconstruct the whole form on every change detection cycle?

**Comment by xealot** [comment]
The example in the plunkr is contrived, but in our real world example we have a component that is reused between routes.

You can navigate from resource to resource from that page (think next/prev record) and the component containing the form never unloads, the data just changes. So when the resource changes we re-initialize the form.

This (or something very similar) was supposedly fixed here:
https://github.com/angular/angular/pull/11051

**Comment by DzmitryShylovich** [comment]
You should use reset https://github.com/angular/angular/pull/9974 with initial value or setValue/patchValue https://github.com/angular/angular/pull/10537

**Comment by xealot** [comment]
Then I have to initialize the form twice, once with blank values and then once with real values. Seems silly, considering I don't think this is supposed to happen.

**Comment by alxhub** [comment]
I think the problem is that `ngOnChanges` runs after change detection, but you are setting `this.form` and expecting change detection to pass it into the `formGroup` directive - which won't happen as CD has already run. The approach itself of recreating the form every change detection cycle is flawed.

**Comment by xealot** [comment]
Then is the blessed way to create a form with a ton of `["", Validators.required]` and then use `reset()` instead?

This is something we can do, it just feels very messy as we're listing all the fields in the form twice.

While I know that there are not going to be breaking changes to the Forms module, I regret that it wasn't designed like many other libs which completely separate form values from validation and structure.

**Comment by xealot** [comment]
I'll also add one last thought... This pattern was specifically addressed in earlier PRs (#11051) but simply seem broken for custom components.
